File: war/lib/War.ex
defmodule War do
  @moduledoc """
    this code plays out the card game war
  """

  @doc """
    Each function has its own documentation but in short:
    the game recursively calls itself until one player is out of card
  """
  def deal(shuf) do
    #starts the game and deals the deck to hand1 and hand2
    #calls the game part and changes the 14s back to 1s
    {hand1, hand2} = dealing(oneToFourteen(shuf))
    fourteenToOne(win(hand1,hand2))
  end

  def dealing(array) do
    #breaks the list into 2 lists for each player
    #adds cards to each player in alternating fashion by comparing the size of the length
    #returns the tuple with both decks
    {list1, list2} = Enum.reduce_while(array, {[], []}, fn x, {list1, list2} ->
    case length(list1) == length(list2) do
      true -> {:cont, {[x | list1], list2}}
      false -> {:cont, {list1, [x | list2]}}
    end
  end)
  {(list1), (list2)}
  end

  def oneToFourteen(list) do
    #changes all the ones to 14 to make it easier to compare in game
    Enum.map(list, fn x -> if x == 1, do: 14, else: x end)
  end

  def fourteenToOne(list) do
    #changes all the 14s to one
    Enum.map(list, fn x -> if x == 14, do: 1, else: x end)
  end

  #recursively calls itself until one player runs out of cards
  #returns the non-empty hand
  def win([], hand2), do: hand2
  def win(hand1, []), do: hand1
  def win(hand1, hand2) do
    {l1, l2} = turn(hand1, hand2) #calls turn to change each players hand
    win(l1, l2) #calls itself recursively
  end

  def turn([h1|t1], [h2|t2]) do
    #checks who won the battle and adds cards to the winner
    #calls warBattle if war occurs
    if (h1 > h2) do
      {t1 ++ [h1, h2], t2}
    else if (h2 > h1) do
      {t1, t2 ++ [h2,h1]}
    else
      warBattle([h1]++t1,[h2]++t2, [])
    end
    end
  end

  #Stays within warBattle until war has ended

  #checks if either player has run out of cards in the middle of war and then
  #returns both hands with the sorted war cards added to the winners hand
  def warBattle([],hand2, holder), do: {[], hand2 ++ Enum.sort(holder, :desc)}
  def warBattle(hand1,[], holder), do: {hand1 ++ Enum.sort(holder, :desc), []}

  #checks for edge cases when the player does not have enough cards to complete a war, so remove cards from both players until one is empty
  def warBattle(hand1, hand2, holder) when (length(hand2) == 2 or length(hand1) == 2) and length(holder)==0, do: warBattle(tl(tl(hand1)), tl(tl(hand2)), holder ++ [hd(hand1)] ++ [hd(hand2)] ++ [hd(tl(hand1))] ++ [hd(tl(hand2))])
  def warBattle(hand1, hand2, holder) when (length(hand1) < 3 or length(hand2) < 3) and length(holder)==0, do: warBattle(tl(hand1), tl(hand2), holder ++ [hd(hand1)] ++ [hd(hand2)])

  #actually plays the war part of the game
  #holder represents all the cards used in the war
  #adds first cards involved in war to holder and keeps adding to holder until war over
  #then adds golder cards to winner
  def warBattle([h1 | t1], [h2 | t2], holder) do
    cards = Enum.sort(holder ++ [h1] ++ [h2], :desc)
    if (h1 == h2) do
      warBattle(tl(t1),tl(t2),Enum.sort(cards ++ [hd(t1)] ++ [hd(t2)], :desc))
    else if (h1 > h2) do
      {t1 ++ cards, t2}
    else
      {t1, t2 ++ cards}
    end
    end
  end
end

File: haskell/lib/War.hs
module War (deal) where

import Data.List

{--
this code plays out the card war game by taking a 52 int array and returning the end game array
--}

--this function gets called by the test file
--the function returns the end og game array recieved from the function win
deal :: [Int] -> [Int]
deal shuf = fourteenToOne(win(hand1, hand2))
    where
        (hand1,hand2) = dealing(oneToFourteen shuf)

--This function changes all the 1s to 14, otherwise returns the same int and returns the array
oneToFourteen :: [Int] -> [Int]
oneToFourteen = map (\x -> if x == 1 then 14 else x)

--This function changes all the 14s to 1, otherwise returns the same int and returns the array
--This function this used to revert the array back to normal before returning
fourteenToOne :: [Int] -> [Int]
fourteenToOne = map (\x -> if x == 14 then 1 else x)

--This function takes one array and returns a tuple with 2 arrays representing hand1 and hand2
dealing :: [Int] -> ([Int], [Int])
dealing array =
  let (list1, list2) = foldl (\(list1, list2) x ->
                                if length list1 == length list2
                                  then (x:list1, list2)
                                  else (list1, x:list2)) ([], []) array
  in (list1, list2)

--this function recursively calls turn until one of the hands run out of cards
--returns the hand that isn't empty
win :: ([Int], [Int]) -> [Int]
win ([],hand2) = hand2
win (hand1,[]) = hand1
win (hand1, hand2) = win nextTurn
    where nextTurn = turn(hand1, hand2) 

--this function is recursively called and changes each players deck based on whether hand1 or hand2 wins
--adds both top cards of each players deck to the winners hand
--if war occurs, the warBattle function is called
turn :: ([Int], [Int]) -> ([Int], [Int])
turn (h1:t1, h2:t2)
  | h1 > h2 = (t1 ++ [h1] ++ [h2], t2)
  | h2 > h1 = (t1, t2 ++ [h2] ++ [h1])
  | otherwise = warBattle (h1:t1, h2:t2, []) 

--this warfunction is called and recursively calls itself until war is over
--holder holds all the cards involved in the war
warBattle :: ([Int], [Int], [Int]) -> ([Int], [Int])
warBattle ([], hand2, holder) = ([], hand2 ++ reverse (sort holder))
warBattle (hand1, [], holder) = (hand1 ++ reverse (sort holder), [])

--this chunk checks for edge cases where the player runs out of cards mid-war
warBattle (hand1, hand2, holder) 
    | (length hand2 < 2 || length hand1 < 2) && length holder < 3 = 
    warBattle(tail hand1, tail hand2, holder ++ [head hand1] ++ [head hand2])

--this chunk of code checks if war has ended and adds the sorted holders array to the winners hand
warBattle (hand1h:hand1t, hand2h:hand2t, holder) = do
  let cards = reverse (sort (holder ++ [hand1h, hand2h]))
  if hand1h == hand2h then warBattle(tail hand1t, tail hand2t, reverse (sort (cards ++ [head hand1t, head hand2t])))
  else if hand1h > hand2h then (hand1t ++ cards, hand2t)
  else (hand1t, hand2t ++ cards)